lib/streamlit/runtime/pages_manager.py

```diff
@@ -15,322 +15,119 @@
-                    # the predicate to circumvent this.
-                    lambda p: p
-                    and (p["url_pathname"] == self.pages_manager.intended_page_name),
-                    self._pages.values(),
-                ),
-                None,
-            )
-
-        return self._pages.get(fallback_page_hash, None)
-
-    def get_pages(self) -> dict[PageHash, PageInfo]:
-        # If pages are not set, provide the common page info where
-        # - the main script path is the executing script to start
-        # - the page script hash and name reflects the intended page requested
-        return self._pages or {
-            self.pages_manager.main_script_hash: {
-                "page_script_hash": self.pages_manager.intended_page_script_hash or "",
-                "page_name": self.pages_manager.intended_page_name or "",
-                "icon": "",
-                "script_path": self.pages_manager.main_script_path,
-            }
-        }
-
-    def set_pages(self, pages: dict[PageHash, PageInfo]) -> None:
-        self._pages = pages
-
-    def register_pages_changed_callback(
-        self,
-        callback: Callable[[str], None],
-    ) -> Callable[[], None]:
-        # V2 strategy does not handle any pages changed event
-        return lambda: None
-
-
 class PagesManager:
-    """
-    PagesManager is responsible for managing the set of pages based on the
-    strategy. By default, PagesManager uses V1 which relies on the original
-    assumption that there exists a `pages` directory with all the scripts.
-
-    If the `pages` are being set directly, the strategy is switched to V2.
-    This indicates someone has written an `st.navigation` call in their app
-    which informs us of the pages.
-
-    NOTE: Each strategy handles its own thread safety when accessing the pages
-    """
-
-    DefaultStrategy: type[PagesStrategyV1 | PagesStrategyV2] = PagesStrategyV1
-
     def __init__(
         self,
         main_script_path: ScriptPath,
-        script_cache: ScriptCache | None = None,
-        **kwargs,
+        *,
+        # setup_watcher is a convenience flag to allow disabling the watcher for testing
+        setup_watcher: bool = True,
     ):
         self._main_script_path = main_script_path
         self._main_script_hash: PageHash = calc_md5(main_script_path)
-        self.pages_strategy = PagesManager.DefaultStrategy(self, **kwargs)
-        self._script_cache = script_cache
-        self._intended_page_script_hash: PageHash | None = None
-        self._intended_page_name: PageName | None = None
         self._current_page_script_hash: PageHash = ""
+        pages_dir = self.main_script_parent / "pages"
+        self._is_mpa_v1 = os.path.exists(pages_dir)
+        if self._is_mpa_v1:
+            if setup_watcher:
+                source_util.setup_pages_watcher(pages_dir)
+            self._pages = source_util.get_pages(self._main_script_path)
```

Comment:
It might be a good idea to do a slight refactoring of `source_util.get_pages` to move the `pages_dir = main_script_path.parent / "pages"` in that function up and return `None` if it not exists. Then this block would become:

```python
self._pages = source_util.get_pages(self._main_script_path)
if self._pages:
  if: setup_watcher:
    source_util.setup_pages_watcher(pages_dir)
else:
  self._pages = {}
```

This way we don't have to deal with the path checking of `pages` etc. multiple times, reducing potential errors where we use the wrong parent path etc.

Reply:
Still technically necessary because new pages being added and removed do not update the set of pages.